security/src/test/groovy/io/micronaut/security/filters/SecurityFilterSpec.groovy

```diff
@@ -0,0 +1,121 @@
+package io.micronaut.security.filters
+
+import io.micronaut.core.convert.value.MutableConvertibleValues
+import io.micronaut.http.HttpMethod
+import io.micronaut.http.HttpRequest
+import io.micronaut.http.HttpResponse
+import io.micronaut.http.filter.ServerFilterChain
+import io.micronaut.security.EmbeddedServerSpecification
+import io.micronaut.security.authentication.Authentication
+import io.micronaut.security.authentication.AuthorizationException
+import io.micronaut.security.config.SecurityConfiguration
+import io.micronaut.security.rules.ReactiveSecurityRule
+import io.micronaut.security.rules.SecurityRule
+import io.micronaut.security.rules.SecurityRuleResult
+import io.micronaut.web.router.RouteMatch
+import io.reactivex.Flowable
+import io.reactivex.schedulers.Schedulers
+import io.reactivex.subscribers.TestSubscriber
+import spock.lang.Unroll
+
+import java.util.concurrent.TimeUnit
+
+import static io.micronaut.security.rules.SecurityRuleResult.ALLOWED
+import static io.micronaut.security.rules.SecurityRuleResult.REJECTED
+import static io.micronaut.security.rules.SecurityRuleResult.UNKNOWN
+
+class SecurityFilterSpec extends EmbeddedServerSpecification {
+
+    @Override
+    String getSpecName() {
+        'SecurityFilterSpec'
+    }
+
+    @Unroll('#description')
+    def "Filter order is correct"() {
+        given:
```

Comment:
In general we try to avoid tests that use mocking. This can be tested by having multiple security rules, both reactive and not that return a value based on a condition (`?ruleA=true`), for example. You can then send multiple requests with different query values to effectively control the behavior of the rules then assert the correct result is returned.

Reply:
You could also just get a reference to the bean and set a local boolean flag

Reply:
I took a stab at a no-mock version. My Groovy/Spock is still really weak as this is the first set of tests I've done this way so I'm not 100% sure if I did everything correctly.